File: codes/taehyeok/525_contiguousarray.py
class Solution(object):
    def findMaxLength(self, nums):
        sum = 0

        dict = {0:-1}
        max_length = 0

        for index, num in enumerate(nums):
            if num == 1:
                sum += 1
            else:
                sum -= 1
            if sum in dict:
                max_length = max(max_length, index - dict[sum])
            else:
                dict[sum] = index
        return max_length

File: codes/naeun/525_contiguousarray.java
class Solution {
    // ex: [0,1,1,1,1,1,0,0,0]->[-1,0,1,2,3,4,3,2,1]
    public int findMaxLength(int[] nums) {
        // 0 -> -1
        // 1 -> 1
        // if sum is zero -> it's match condition
        Map<Integer, Integer> sumIndex = new HashMap<>();
        int sum = 0;
        int answer = 0;
        for(int i = 0 ; i < nums.length; i++){
            if(nums[i] == 0){
                sum--;
            } else {
                sum++;
            }
            if(sum == 0){
                answer = i+1;
            }
            sumIndex.put(sum, i);
        }

        sum = 0;
        for(int left = 0; left < nums.length; left++){
            if(nums[left] == 0){
                sum--;
            } else {
                sum++;
            }

            answer = Math.max(answer, sumIndex.get(sum) - left);
        }

        return answer;
    }
}